File: examples/client-gen/tictactoe/instructions/setup_game.py
from __future__ import annotations
import typing
from solders.pubkey import Pubkey
from solders.system_program import ID as SYS_PROGRAM_ID
from solders.instruction import Instruction, AccountMeta
from anchorpy.borsh_extension import BorshPubkey
import borsh_construct as borsh
from ..program_id import PROGRAM_ID


class SetupGameArgs(typing.TypedDict):
    player_two: Pubkey


layout = borsh.CStruct("player_two" / BorshPubkey)


class SetupGameAccounts(typing.TypedDict):
    game: Pubkey
    player_one: Pubkey


def setup_game(
    args: SetupGameArgs,
    accounts: SetupGameAccounts,
    program_id: Pubkey = PROGRAM_ID,
    remaining_accounts: typing.Optional[typing.List[AccountMeta]] = None,
) -> Instruction:
    keys: list[AccountMeta] = [
        AccountMeta(pubkey=accounts["game"], is_signer=True, is_writable=True),
        AccountMeta(pubkey=accounts["player_one"], is_signer=True, is_writable=True),
        AccountMeta(pubkey=SYS_PROGRAM_ID, is_signer=False, is_writable=False),
    ]
    if remaining_accounts is not None:
        keys += remaining_accounts
    identifier = b"\xb4\xda\x80K:\xde#R"
    encoded_args = layout.build(
        {
            "player_two": args["player_two"],
        }
    )
    data = identifier + encoded_args
    return Instruction(program_id, data, keys)

File: ts-reference/examples/tic-tac-toe/generated-client/instructions/setupGame.ts
import { TransactionInstruction, PublicKey } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface SetupGameArgs {
  playerTwo: PublicKey
}

export interface SetupGameAccounts {
  game: PublicKey
  playerOne: PublicKey
  systemProgram: PublicKey
}

export const layout = borsh.struct([borsh.publicKey("playerTwo")])

export function setupGame(args: SetupGameArgs, accounts: SetupGameAccounts) {
  const keys = [
    { pubkey: accounts.game, isSigner: true, isWritable: true },
    { pubkey: accounts.playerOne, isSigner: true, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([180, 218, 128, 75, 58, 222, 35, 82])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      playerTwo: args.playerTwo,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}
